analyzers/src/SonarAnalyzer.CSharp/Rules/ExpectedExceptionAttributeShouldNotBeUsed.cs

```diff
@@ -18,18 +18,51 @@
  * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
  */
 
+using SonarAnalyzer.Common.Walkers;
+
 namespace SonarAnalyzer.Rules.CSharp
 {
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     public sealed class ExpectedExceptionAttributeShouldNotBeUsed : ExpectedExceptionAttributeShouldNotBeUsedBase<SyntaxKind>
     {
         protected override ILanguageFacade<SyntaxKind> Language => CSharpFacade.Instance;
 
-        protected override bool HasMultiLineBody(SyntaxNode syntax)
+        protected override bool HasMultiLineBody(SyntaxNode node)
         {
-            var declaration = (MethodDeclarationSyntax)syntax;
+            var declaration = (MethodDeclarationSyntax)node;
             return declaration.ExpressionBody is null
                 && declaration.Body?.Statements.Count > 1;
         }
+
+        protected override bool AssertInCatchFinallyBlock(SyntaxNode node)
+        {
+            var walker = new CatchFinallyAssertion();
+            foreach (var x in node.DescendantNodes(_ => true).Where(x => x.IsAnyKind(SyntaxKind.CatchClause, SyntaxKind.FinallyClause)))
+            {
+                walker.SafeVisit(x);
+            }
+            return walker.HasAssertion;
+        }
+
+        private sealed class CatchFinallyAssertion : SafeCSharpSyntaxWalker
+        {
+            public bool HasAssertion { get; set; }
+
+            public override void VisitInvocationExpression(InvocationExpressionSyntax node)
+            {
+                HasAssertion = node.Expression
+                    .ToString()
+                    .SplitCamelCaseToWords()
+                    .Intersect(UnitTestHelper.KnownAssertionMethodParts)
+                    .Any();
```

Comment:
This may not cover all Assertion method.
It does not test as much as in [TestMethodShouldContainAssertion](https://github.com/SonarSource/sonar-dotnet/blob/master/analyzers/src/SonarAnalyzer.CSharp/Rules/TestMethodShouldContainAssertion.cs#L90-L124).

I will create a ticket to have a common place to test for assertion methods.

Reply:
#9619